servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -194,53 +197,30 @@ public void onSubscribe(final Subscription s) {
         public void onNext(@Nullable final T t) {
             final Publisher<? extends R> publisher = requireNonNull(source.mapper.apply(t));
             FlatMapPublisherSubscriber<T, R> subscriber = new FlatMapPublisherSubscriber<>(this);
-            if (!cancellableSubscribers.add(subscriber)) {
-                return;
-            }
-            for (;;) {
-                final int prevActiveSources = activeMappedSources;
-                if (prevActiveSources < 0) {
-                    // We have been cancelled, or already completed and the active count flipped to negative, either
-                    // way we don't want to Subscribe or retain a reference to this Publisher.
-                    cancellableSubscribers.remove(subscriber);
-                    break;
-                } else if (activeMappedSourcesUpdater.compareAndSet(this, prevActiveSources, prevActiveSources + 1)) {
-                    publisher.subscribeInternal(subscriber);
-                    break;
-                }
+            if (cancellableSubscribers.add(subscriber) && activeMappedSourcesUpdater.incrementAndGet(this) > 0) {
+                publisher.subscribeInternal(subscriber);
             }
+            // else if activeMapped <=0 after increment we have already terminated and onNext isn't valid!
         }
 
         @Override
         public void onError(final Throwable t) {
-            if (!onError0(t, true)) {
-                LOGGER.debug("Already terminated/cancelled, ignoring error notification.", t);
+            try {
+                upstreamError = t;
+                enqueueAndDrain(error(t));
+            } finally {
+                doCancel(false, true);
```

Comment:
We should cancel before draining signals. Consider adding a comment explaining why "cancel-before-onError" is important in cases when we use this pattern:
```java
try {
    // cancel
} finally {
    // onError
}
```

Reply:
in this case we aren't cancelling the upstream subscription because we have been terminated by upstream. However I'll make it consistent with flatMapMergeSingle.